feat(ast): add loc field support for AST nodes

mertcanaltin opened the pull request:
Add line and column information to AST serialization:

extend UTF8ToUtf16 converter with line tracking
support all line break types (\n, \r, \r\n, LS, PS)
add loc options to ESTree serializer configs
implement efficient offset-to-line-column conversion
add SourceLocation struct for ESTree compatibility
integrate with NAPI parser bindings

fyi: #10307

overlookmotel commented on crates/oxc_ast_visit/src/utf8_to_utf16/translation.rs:
```diff
@@ -75,13 +87,79 @@ impl AlignedChunk {
+            // Handle line breaks
+            if track_lines {
+                let mut is_line_break = false;
+                let mut line_break_len = 1;
+
+                match byte {
+                    b'\n' => is_line_break = true,
+                    b'\r' => {
+                        is_line_break = true;
+                        // Check for \r\n
+                        if index + 1 < slice.len() && slice[index + 1] == b'\n' {
+                            line_break_len = 2;
+                        }
+                    }
+                    // Unicode line separators LS (\u2028) and PS (\u2029)
+                    // LS: E2 80 A8, PS: E2 80 A9
+                    0xE2 if index + 2 < slice.len()
+                        && slice[index + 1] == 0x80
+                        && (slice[index + 2] == 0xA8 || slice[index + 2] == 0xA9) =>
+                    {
+                        is_line_break = true;
+                        line_break_len = 3;
+                    }
+                    _ => {}
+                }
+
+                if is_line_break {
+                    let line_end_offset = start_offset + index + line_break_len;
+                    if line_end_offset < source_text.len() {
+                        #[expect(clippy::cast_possible_truncation)]
+                        let utf8_offset = line_end_offset as u32;
+                        lines
+                            .as_mut()
+                            .unwrap()
+                            .push(LineTranslation { utf8_offset, utf16_difference });
+                    }
+                    index += line_break_len;
+                    continue;
+                }
+            }
```
I can see a couple of problems here.

1. `process_slice` is only called in the main search loop (`while ptr < body_end_ptr { ... }`) if the chunk contains non-ASCII characters. So line breaks won't be recorded unless they're at the very start or end of the source text, or they happen to have a non-ASCII char close to them.

The `chunk.contains_unicode()` check in that loop needs to be replaced with a check for "chunk contains unicode OR `\r` / `\n`".

2. Irregular line breaks are non-ASCII, so need `Translation` records as well as a `LineTranslation` record. Simplest way to do that is skip the `continue` statement if `line_break_len == 3`.

Also, a small optimization: No need for 2 vars `is_line_break` and `line_break_len`. Just `line_break_len` would do, with `0` representing "not a line break".

overlookmotel reviewed (commented):
Obviously this is incomplete at present - nothing actually calls `serialize_span_with_loc` at the moment, so the JSON won't include `loc`s (as per the TODO comments).

Please let me know if you need any help getting the translation table into the serializer.

I've made a couple of comments in the meantime - apologies if they're premature, and you just haven't got to this yet.

One other broader point: I don't think we need separate `*_with_loc` methods. `serialize_span_with_loc` shouldn't be needed. Ditto `new_with_loc`, and the extra `to_estree_ts_json_with_loc` etc methods. The existing methods can just have an additional `loc: bool` param added.

We're not at 1.0 yet, and regularly make breaking changes. Please don't worry about changing existing APIs.

I've marked the PR as draft for now.

mertcanaltin commented:
@overlookmotel You were spot on, fixed everything:

  - Line break detection now works throughout text (not just near Unicode)
  - Added chunk boundary fallback for PS/LS edge cases
  - Removed redundant `*_with_loc` methods, added `loc` param instead
  - `serialize_span` now auto-translates, showing real coordinates not (0,0)

mertcanaltin commented on crates/oxc_ast/src/serialize/mod.rs:
```diff
@@ -196,17 +236,17 @@ fn get_ts_start_span(program: &Program<'_>) -> u32 {
-                && let Some(decorator) = class.decorators.first()
-            {
-                return cmp::min(start, decorator.span.start);
+            if let Some(Declaration::ClassDeclaration(class)) = &decl.declaration {
+                if let Some(decorator) = class.decorators.first() {
+                    return cmp::min(start, decorator.span.start);
+                }
```
Thanks, I edited

overlookmotel replied:
Looks like these changes are still present.

mertcanaltin replied:
thanks, I missed this topic

overlookmotel commented on crates/oxc_ast_visit/src/utf8_to_utf16/translation.rs:
```diff
@@ -75,13 +98,97 @@ impl AlignedChunk {
+
+                if line_break_len > 0 {
+                    let line_end_offset = start_offset + index + line_break_len;
+                    if line_end_offset < source_text.len() {
```
I don't think we should have this check. If file ends with a line break, I think need a `LineTranslation` for it, so that converting offset 2 for source `x\n` gives you `line: 2, column: 0`, not `line: 1, column: 2`.

mertcanaltin replied:
thanks, I edited

overlookmotel commented on crates/oxc_ast_visit/src/utf8_to_utf16/translation.rs:
```diff
@@ -150,10 +259,16 @@ pub fn build_translations(source_text: &str, translations: &mut Vec<Translation>
-            // `ptr` is derived from `start_ptr`.
-            let offset = unsafe { ptr.offset_from_unsigned(start_ptr) };
+
+        // SAFETY: `ptr` is equal to or after `start_ptr`. Both are within bounds of `bytes`.
+        // `ptr` is derived from `start_ptr`.
+        let offset = unsafe { ptr.offset_from_unsigned(start_ptr) };
```
Move this into the `if has_unicode || has_line_breaks { ... }` block. If no line breaks or unicode chars in the block, we don't use this value, so avoid calculating it in that case.

overlookmotel commented on crates/oxc_ast_visit/tests/test_loc_integration.rs:
```diff
@@ -0,0 +1,52 @@
+use oxc_ast_visit::utf8_to_utf16::Utf8ToUtf16;
+
+#[test]
+fn test_loc_integration() {
```
Can all these tests move into `src/mod.rs`, with the rest of the tests please?

overlookmotel reviewed (commented):
I've pushed a couple of fixes just to get more of it to compile.

But still some compilation failures, and some tests are failing.

All of these should pass:

```rs
cargo test --all-features
just test-estree
cd napi/parser; pnpm run build-test; RUN_RAW_TESTS=true pnpm test; cd ../..
```

I've made a few comments. I don't know if they explain the test failures or not.

Just to say (and I hope you don't consider this rude): Obviously we all rely on AI these days, but in my opinion our would-be robot overlords are not quite up to the task yet. So they need a bit of human hand-holding and checking of their work.

Could you please make sure the code compiles and tests pass before next round of review?

Of course, if you run into problems or want to discuss anything, very happy to help.

mertcanaltin committed:
feat(ast): add loc field support for AST nodes

mertcanaltin committed:
added convert_offset_back & convert_span_back method

mertcanaltin committed:
fix: address review feedback for loc field implementation

mertcanaltin committed:
fix: address review feedback for loc field implementation

mertcanaltin committed:
fix: update method calls for new loc parameter API

mertcanaltin committed:
fix: resolve unstable let expressions in request.rs

mertcanaltin commented:
Hello, I thought these warnings weren't related to my PR. Would you mind commenting on this? @overlookmotel 

```
➜  oxc git:(mert/feat/add-loc-field-support) cargo ck                                       
    Checking oxc_ast v0.87.0 (/Users/mertcanaltin/Desktop/projects/oxc/crates/oxc_ast)
    Checking oxc_cfg v0.87.0 (/Users/mertcanaltin/Desktop/projects/oxc/crates/oxc_cfg)
    Checking rustls-webpki v0.103.4
    Checking tempfile v3.21.0
    Checking criterion2 v3.0.2
    Checking oxc_linter_codegen v0.0.0 (/Users/mertcanaltin/Desktop/projects/oxc/tasks/linter_codegen)
    Checking oxc_ast_tools v0.0.0 (/Users/mertcanaltin/Desktop/projects/oxc/tasks/ast_tools)
    Checking oxc_regular_expression v0.87.0 (/Users/mertcanaltin/Desktop/projects/oxc/crates/oxc_regular_expression)
error[E0658]: `let` expressions in this position are unstable
  --> tasks/linter_codegen/src/main.rs:48:12
   |
48 |         if let Some(src_path) = find_rule_source_file(&root, rule)
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
  --> tasks/linter_codegen/src/main.rs:49:16
   |
49 |             && let Ok(src_contents) = fs::read_to_string(&src_path)
   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
  --> tasks/linter_codegen/src/main.rs:50:16
   |
50 |             && let Ok(file) = syn::parse_file(&src_contents)
   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
  --> tasks/linter_codegen/src/main.rs:51:16
   |
51 |             && let Some(node_types) = detect_top_level_node_types(&file, rule)
   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
  --> tasks/linter_codegen/src/main.rs:92:8
   |
92 |     if let Some(node_types) = node_types
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
  --> tasks/linter_codegen/src/main.rs:99:8
   |
99 |     if let Some(node_types) = node_types
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
 --> tasks/linter_codegen/src/utils.rs:4:8
  |
4 |     if let Expr::MethodCall(mc) = expr
  |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
 --> tasks/linter_codegen/src/utils.rs:7:12
  |
7 |         && let Expr::Path(p) = &*mc.receiver
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

For more information about this error, try `rustc --explain E0658`.
error: could not compile `oxc_linter_codegen` (bin "oxc_linter_codegen") due to 8 previous errors
warning: build failed, waiting for other jobs to finish...
error: could not compile `oxc_linter_codegen` (bin "oxc_linter_codegen" test) due to 8 previous errors
error[E0658]: `let` expressions in this position are unstable
   --> tasks/ast_tools/src/generators/assert_layouts.rs:241:16
    |
241 |             if let Some(field_niche) = &field_layout.niche
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
   --> tasks/ast_tools/src/generators/raw_transfer_lazy.rs:348:20
    |
348 |                 if let TypeDef::Struct(struct_def) = type_def
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
   --> tasks/ast_tools/src/generators/raw_transfer_lazy.rs:366:16
    |
366 |             if let TypeDef::Struct(struct_def) = type_def
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
   --> tasks/ast_tools/src/generators/scopes_collector.rs:332:16
    |
332 |             if let TypeDef::Cell(cell_def) = field_type
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
   --> tasks/ast_tools/src/generators/scopes_collector.rs:333:20
    |
333 |                 && let TypeDef::Option(option_def) = cell_def.inner_type(schema)
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
   --> tasks/ast_tools/src/generators/typescript.rs:102:8
    |
102 |     if let Some(converter_name) = &struct_def.estree.via
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
   --> tasks/ast_tools/src/generators/typescript.rs:251:12
    |
251 |         if let TypeDef::Struct(field_type) = field.type_def(schema)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
   --> tasks/ast_tools/src/generators/typescript.rs:252:16
    |
252 |             && let Some(flatten_field) = get_single_field(field_type, schema)
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
   --> tasks/ast_tools/src/generators/typescript.rs:312:8
    |
312 |     if let Some(converter_name) = &enum_def.estree.via
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
   --> tasks/ast_tools/src/generators/typescript.rs:351:23
    |
351 |             } else if let Some(converter_name) = &struct_def.estree.via
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
   --> tasks/ast_tools/src/generators/typescript.rs:352:20
    |
352 |                 && let Some(type_name) = get_ts_type_for_converter(converter_name, schema)
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
   --> tasks/ast_tools/src/generators/typescript.rs:362:23
    |
362 |             } else if let Some(converter_name) = &enum_def.estree.via
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
   --> tasks/ast_tools/src/generators/typescript.rs:363:20
    |
363 |                 && let Some(type_name) = get_ts_type_for_converter(converter_name, schema)
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
   --> tasks/ast_tools/src/generators/typescript.rs:423:8
    |
423 |     if let Some(field) = fields_which_are_not_skipped.next()
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
   --> tasks/ast_tools/src/generators/visit.rs:494:12
    |
494 |         if let Some((exit_index, _)) = scope_exit
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
   --> tasks/ast_tools/src/generators/visit.rs:502:12
    |
502 |         if let Some((enter_index, _)) = scope_entry
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
  --> tasks/ast_tools/src/output/mod.rs:92:8
   |
92 |     if let Ok(existing_data) = fs::read(path)
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
   --> tasks/ast_tools/src/parse/load.rs:205:12
    |
205 |         if let Meta::NameValue(name_value) = meta
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
   --> tasks/ast_tools/src/parse/parse.rs:485:16
    |
485 |             if let Meta::NameValue(name_value) = &attr.meta
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
   --> tasks/ast_tools/src/parse/parse.rs:487:20
    |
487 |                 && let Expr::Lit(expr_lit) = &name_value.value
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error[E0658]: `let` expressions in this position are unstable
   --> tasks/ast_tools/src/parse/parse.rs:488:20
    |
488 |                 && let Lit::Str(lit_str) = &expr_lit.lit
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information

error: could not compile `oxc_ast_tools` (bin "oxc_ast_tools") due to 21 previous errors
error: could not compile `oxc_ast_tools` (bin "oxc_ast_tools" test) due to 21 previous errors
➜  oxc git:(mert/feat/add-loc-field-support) 
```

overlookmotel commented:
> Hello, I thought these warnings weren't related to my PR. Would you mind commenting on this? @overlookmotel
> 
> ```
> error[E0658]: `let` expressions in this position are unstable
>   --> tasks/linter_codegen/src/main.rs:48:12
>    |
> 48 |         if let Some(src_path) = find_rule_source_file(&root, rule)
>    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> ```

Are you using an older Rust version on your local machine? Oxc's minimum supported Rust version is 1.87.0.

https://github.com/oxc-project/oxc/blob/73460999aeb007dd43e218b843d54bc8aa058bb1/Cargo.toml#L16-L18

"let chains" only became stable in 1.87.0, so I'd expect the kind of errors you're seeing if on an earlier Rust version.

mertcanaltin committed:
repair

mertcanaltin committed:
repair

mertcanaltin committed:
added test for loc

mertcanaltin committed:
ts repair for loc

mertcanaltin committed:
ts repair for loc

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

mertcanaltin committed:
changes for review comment

mertcanaltin committed:
changes for review comment

mertcanaltin committed:
changes for review comment

mertcanaltin committed:
changes for review comment

mertcanaltin committed:
changes for review comment

mertcanaltin committed:
changes for review comment

mertcanaltin committed:
changes for review comment

mertcanaltin committed:
changes for review comment

mertcanaltin committed:
changes for review comment

mertcanaltin committed:
changes for review comment

mertcanaltin committed:
conflict solved

mertcanaltin committed:
conflict solved

mertcanaltin committed:
conflict solved

mertcanaltin committed:
conflict solved

mertcanaltin committed:
fix: correct comments and assertions in UTF-8 to UTF-16 conversion logic

mertcanaltin committed:
fix: add SourceLocation and loc field to generated types.d.ts

mertcanaltin committed:
fix: use let chains to fix clippy collapsible_if warning

mertcanaltin commented:
Hey @overlookmotel, I've pushed the latest fixes

Fixed collapsible_if clippy warning using let chains

Added SourceLocation interface and [loc](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) field to the TypeScript generator so it's properly included in generated types.d.ts

Fixed the uft8_offset typo
Updated assertion to [full_offset + 3 <= source_text.len()](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) for PS/LS detection

Note: The large number of file changes is because I ran just ast to regenerate the TypeScript type definitions. The actual logic change is minimal - just adding SourceLocation and [loc](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) to the Span interface in the generator.

Waiting for CI to pass. Let me know if anything else needs attention!

mertcanaltin committed:
fix: add safety comment for unsafe block in converter.rs

mertcanaltin commented:
I'm seeing some issues that need clarification:

The just ast / cargo run -p oxc_ast_tools command fails to format JS/TS files because oxfmt panics with "External formatter must be set when napi feature is enabled". This causes the AST Changes CI to fail due to formatting differences.
I see that convert_offset_back and convert_span_back methods are mentioned as missing - I'll need to restore these from the merge conflict resolution.

Could you advise on the best way to proceed with the oxfmt formatting issue?

mertcanaltin committed:
Merge branch 'main' into mert/feat/add-loc-field-support

mertcanaltin committed:
feat(napi): add loc field to Comment struct and integrate loc provider

mertcanaltin committed:
fix(napi):correct loc test assertions and fix invalid test cases

mertcanaltin committed:
Merge branch 'main' into mert/feat/add-loc-field-support

mertcanaltin committed:
perf(napi): use linear cursorfor loc offset lookups during serialization

mertcanaltin committed:
Merge branch 'main' into mert/feat/add-loc-field-support

mertcanaltin committed:
Merge branch 'main' into mert/feat/add-loc-field-support

mertcanaltin committed:
Merge branch 'main' into mert/feat/add-loc-field-support

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

AugustinMauroy commented:
any news ?

mertcanaltin commented:
I continue this to this pr https://github.com/oxc-project/oxc/pull/22774

mertcanaltin commented:
Superseded by #22774, which replays the foundation cleanly on current main as a small, reviewable PR.

This branch had diverged from main by ~1849 commits and the history accumulated noise that wasn't worth rebasing through. The new PR addresses @overlookmotel's earlier feedback about keeping changes small; ESTree `Config`, serializer integration, and the NAPI `loc` option will follow as separate PRs on top of it.
